### PR TITLE
reworked notebook tests

### PR DESCRIPTION
**Pull request by aweng98**
Refactored notebook test suites to make tests more efficient.
- Reduced number of test files by eliminating redundant steps (coverage is not reduced). 
- Reduced number of new workspaces needed.
- Increased test coverage.

`notebook-create.spec` (combined with deleted `notebook-download.spec`, `notebook-r.spec` and `notebook-python3.spec` tests):
- 'Run Python code and download notebook'
- 'Run R code'

`notebook-owner-actions.spec`:
- 'Notebook name must be unique in same workspace'
- 'Duplicate notebook'
- 'Rename notebook'

`notebook-reader-actions.spec`:
- 'Share notebook to workspace READER'
- Workspace READER copy notebook to another workspace'
- 'Workspace READER edit copy of notebook in workspace clone'

### CODE REVIEW

File: e2e/app/page/notebook-preview-page.ts [hunk 1]
```diff
@@ -40,9 +39,25 @@ export default class NotebookPreviewPage extends AuthenticatedPage {
...
-    return getPropValue<string>(codeContent, 'textContent');
+  getEditLink(): Link {
+    return new Link(this.page, Selector.editLink);
+  }
+
+  getRunPlaygroundModeLink(): Link {
+    return new Link(this.page, Selector.runPlaygroundModeLink);
+  }
+
+  async getFormattedCode(): Promise<string[]> {
```
Comment (aweng98) [inline e2e/app/page/notebook-preview-page.ts:42]: Function returns array instead of a glob of string for more precise string check.

**Review by petesantos — APPROVED** [review]
LGTM, minor comments.
Ssorry for the delay on the PR!

File: e2e/tests/datasets/export-to-notebook.spec.ts [hunk 2]
```diff
@@ -123,11 +123,23 @@ describe('Export to notebook tests', () => {
     const code = await notebookPreviewPage.getFormattedCode();
     switch (kernelLanguage.LANGUAGE) {
       case Language.Python:
-        expect(code).toContain('import pandas');
-        expect(code).toContain('import os');
+        expect(
+          code.some((item) => {
+            return item.includes('import pandas');
+          })
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/tests/datasets/export-to-notebook.spec.ts:129]: 👍🏻 Cool. New version is shorter.

File: e2e/app/page/notebook-preview-page.ts [hunk 3]
```diff
@@ -40,9 +39,25 @@ export default class NotebookPreviewPage extends AuthenticatedPage {
     return notebookPage;
   }
 
-  async getFormattedCode(): Promise<string> {
-    const codeContent = await this.waitForCssSelector('#notebook-container pre');
-    return getPropValue<string>(codeContent, 'textContent');
+  async getFormattedCode(): Promise<string[]> {
+    const css = '#notebook-container pre';
+    await this.waitForCssSelector(css);
+    const textContents = await this.findNotebookIframe().then((frame) => {
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/app/page/notebook-preview-page.ts:45]: No. sometimes there are more than 1 frame. So finding and picking the right frame is needed.

File: e2e/app/page/notebook-preview-page.ts [hunk 4]
```diff
@@ -40,9 +39,25 @@ export default class NotebookPreviewPage extends AuthenticatedPage {
...
+      return frame.$$(css);
+    });
+    const arr: string[] = [];
+    for (const textContent of textContents) {
+      arr.push(await getPropValue<string>(textContent, 'textContent'));
+    }
+    return arr;
```
[the remark this answers is not in the document]
Comment (aweng98) [reply e2e/app/page/notebook-preview-page.ts:52]: Using `fp.flow`, eslint complained about the wrong return type.

This code compiles
```
return Promise.all(textContents.map(async (textContent) => await getPropValue<string>(textContent, 'textContent')));
```

**Review by aweng98 — COMMENTED** [review]
Thank you. No problem with the wait, this was not an urgent PR.

**Pull request merged**